File: CS333 - Programming Languages/Syntax/C++/task3.cpp
/*
 * File: task3.cpp
 * Purpose: To show all the different data types in C++
 * Author: Samuel Munoz
 * Date: 03-03-2020
 */

#include <iostream>

class Person {
	public:
		Person(char *n) { name = n; }
		char * getName() { return name; }
		void setName(char *n) { name = n; } 
	private:
		char *name;
};

typedef struct {
	int num;
} Container;

enum Comparsion {
	Less,
	Greater,
	Equal
};

// unions are a way to create different 
typedef union {
	bool b;
	int i;
} Optional;

void printHelloWorld(); // derivated type
void printHelloWorld() { std::cout << "Hello World!" << std::endl; }
void printArray(int *head, int size) {
	std::cout << "[";
	for(int i=0;i<size;i++) {
		if(i != size-1)
			std::cout << head[i] << ", ";
		else
			std::cout << head[i] << "]" << std::endl;
	}
}

int main(int argc, char **argv) {
	
	// All primary data types
	// integers
	int i = 1; 
	// characters
	char c = 'a';
	// booleans
	bool b = false;
	// floats
	float f = 3.14;
	// doubles
	double d = 1.69;
	// wide char; has more bytes to expand the number of different chars inside the variable. Bigger than unsigned char
	wchar_t w = L'w';

	// Derivated types
	// functions
	int array[10] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 }; // array (can be of any type)
	int *ptr; // pointer (can be of any type)
	// reference vs pointers
	// references must refer to a specific location of memory
	// references must be initialized when references are delcared
	// reference's values cannot be changed after a reference is created
	int &ref = array[0]; // reference (can be of any type)

	// User Defined Types
	char str[] = "Juan";
	// class
	Person p = Person(str);
	// struct
	Container container;
	container.num = 5;
	// union
	Optional o;
	o.b = false;
	// enumeration
	Comparsion en;

	std::cout << "Primary Data Types" << std::endl;
	std::cout << "Boolean\t\t\t" << "size: " << sizeof(bool) << "\t\tvalue: " << b << "\t\t!: " << !b << std::endl;
	std::cout << "Integer\t\t\tsize: " << sizeof(int) << "\t\tvalue: " << i << "\t\t+: " << i+1 << "\t-: " << i-1 << "\t*: " << i*2 << "\t/: " << i/2 << "\t%: " << i%2 << std::endl;
	std::cout << "Character\t\tsize: " << sizeof(char) << "\t\tvalue: " << c << "\t\t+: " << c+1 << "\t-: " << c-1 << "\t*: " << c*2 << "\t/: " << c/2 << "\t%: " << c%2 << std::endl;
	std::cout << "Float\t\t\tsize: " << sizeof(float) << "\t\tvalue: " << f << "\t\t+: " << f+1 << "\t-: " << f-1 << "\t*: " << f*2 << "\t/: " << f/2 << std::endl;
	std::cout << "Double\t\t\tsize: " << sizeof(int) << "\t\tvalue: " << d << "\t\t+: " << d+1 << "\t-: " << d-1 << "\t*: " << d*2 << "\t/: " << d/2 << std::endl;
	std::cout << "Wide Character\t\tsize: " << sizeof(wchar_t) << "\t\tvalue: " << w << "\t\t+: " << w+1 << "\t-: " << w-1 << "\t*: " << w*2 << "\t/: " << w/2 << "\t%: " << w%2 << std::endl;

	std::cout << "\n\nDerived Data Types" << std::endl;

	std::cout << "Function\t\t";
	printHelloWorld();
	std::cout << "Array\t\t\t";
	printArray(array, 10);
	std::cout << "Pointer\t\t\t" << ptr << std::endl;
	std::cout << "Reference\t\t" << ref << std::endl;
	
	std::cout << "\n\nUser-Defined Data Types" << std::endl;

	std::cout << "Classes\t\t\tName: " << p.getName() << std::endl;
	std::cout << "Structures\t\tNumber: " << container.num << std::endl;
	std::cout << "Union:" << std::endl;
	std::cout << "\t\t\t(Integer Corrupted)" << "\tBoolean: " << o.b << "\t\tInteger: " << o.i << std::endl;
	o.i = 34;
	std::cout << "\t\t\t(Boolean Corrupted)" << "\tBoolean: " << o.b << "\t\tInteger: " << o.i << std::endl;
	std::cout << "Enumeration\t\t" << "Less: " << (en=Less) << "\t\t\tGreater: " << (en=Greater) << "\t\tEqual: " << (en=Equal) << std::endl;

	return 0;
}

/*
Sources
https://www.geeksforgeeks.org/c-data-types/
https://www.tutorialspoint.com/cplusplus/cpp_references.htm
https://www.tutorialspoint.com/cprogramming/c_unions.htm
https://www.programiz.com/cpp-programming/enumeration
*/

File: CS333 - Programming Languages/Syntax/C#/task3.cs
/*
 * File: Program.cs
 * Purpose: To show all the different data types in C#
 * Author: Samuel Munoz
 * Date: 03-03-2020
 */

using System;

namespace task3
{
	interface NotUsed {
		void printNotUsed();
	}

	class Person {
		private string name;

		public Person(string s) { name = s; } 
		public string getName() { return name; }
		public void setName(string s) { name = s; }
	}

    class Program
    {
        static void Main(string[] args)
        {
            // boolean value
            bool boolean = true;
            // 8-bit unsigned integer
            byte b = 1;
            // 16-bit Unicode character
            char c = 'a';
            // 128-bit precise decimal values with 28-29 significant digits
            decimal dec = 2;
            // 64-bit double-precision floating point type
            double dou = 3;
            // 32-bit single-precision floating point type
            float f = 4;
            // 32-bit signed integer type
            int i = 5;
            // 64-bit signed integer type
            long l = 6;
            // sbyte: 8-bit signed integer type
            sbyte sb = 7;
            // 16-bit signed integer type
            short s = 8;
            // 32-bit unsigned integer type
            uint ui = 9;
            // 64-bit unsigned integer type
            ulong ul = 10;
		// 16-bit unsigned integer type
            ushort us = 11;

            // Object type: object is base class for all other objects
            object o = 100; // o is a reference to the object created
            // Dynamic type: it is placeholder that can changes type as variable's type is determined during runtime
            dynamic d_int = 20;
            dynamic d_char = 'd';
            // String types
            string str = "New string";
            // Pointers
            // int *ptr = &i; Cannot run since pointers can only be runned in unsafe context
            // Classes
            Person p = new Person("Juan");

            Console.WriteLine("Value Types");
            Console.WriteLine("Boolean:\tsize: " + sizeof(bool) + "\t\tvalue: " + boolean + "\t\t!: " + !boolean);
            Console.WriteLine("Byte:\t\tsize: " + sizeof(byte) + "\t\tvalue: " + b + "\t\t+: " + (b+1) + "\t-: " + (b-1) + "\t*: " + (b*2) + "\t/: " + (b/2) + "\t%: " + (b%2));
            Console.WriteLine("Character:\tsize: " + sizeof(char) + "\t\tvalue: " + c + "\t\t+: " + (c+1) + "\t-: " + (c-1) + "\t*: " + (c*2) + "\t/: " + (c/2) + "\t%: " + (c%2));
            Console.WriteLine("Decimal:\tsize: " + sizeof(decimal) + "\tvalue: " + dec + "\t\t+: " + (dec+1) + "\t-: " + (dec-1) + "\t*: " + (dec*2) + "\t/: " + (dec/2) + "\t%: " + (dec%2));
            Console.WriteLine("Double:\t\tsize: " + sizeof(double) + "\t\tvalue: " + dou + "\t\t+: " + (dou+1) + "\t-: " + (dou-1) + "\t*: " + (dou*2) + "\t/: " + (dou/2) + "\t%: " + (dou%2));
            Console.WriteLine("Float:\t\tsize: " + sizeof(float) + "\t\tvalue: " + f + "\t\t+: " + (f+1) + "\t-: " + (f-1) + "\t*: " + (f*2) + "\t/: " + (f/2) + "\t%: " + (f%2));
            Console.WriteLine("Integer:\tsize: " + sizeof(int) + "\t\tvalue: " + i + "\t\t+: " + (i+1) + "\t-: " + (i-1) + "\t*: " + (i*2) + "\t/: " + (i/2) + "\t%: " + (i%2));
            Console.WriteLine("Long:\t\tsize: " + sizeof(long) + "\t\tvalue: " + l + "\t\t+: " + (l+1) + "\t-: " + (l-1) + "\t*: " + (l*2) + "\t/: " + (l/2) + "\t%: " + (l%2));
            Console.WriteLine("Sbyte:\t\tsize: " + sizeof(sbyte) + "\t\tvalue: " + sb + "\t\t+: " + (sb+1) + "\t-: " + (sb-1) + "\t*: " + (sb*2) + "\t/: " + (sb/2) + "\t%: " + (sb%2));
            Console.WriteLine("Short:\t\tsize: " + sizeof(short) + "\t\tvalue: " + s + "\t\t+: " + (s+1) + "\t-: " + (s-1) + "\t*: " + (s*2) + "\t/: " + (s/2) + "\t%: " + (s%2));
            Console.WriteLine("Uint:\t\tsize: " + sizeof(uint) + "\t\tvalue: " + ui + "\t\t+: " + (ui+1) + "\t-: " + (ui-1) + "\t*: " + (ui*2) + "\t/: " + (ui/2) + "\t%: " + (ui%2));
            Console.WriteLine("Ulong:\t\tsize: " + sizeof(ulong) + "\t\tvalue: " + ul + "\t\t+: " + (ul+1) + "\t-: " + (ul-1) + "\t*: " + (ul*2) + "\t/: " + (ul/2) + "\t%: " + (ul%2));
            Console.WriteLine("Ushort:\t\tsize: " + sizeof(ushort) + "\t\tvalue: " + us + "\t\t+: " + (us+1) + "\t-: " + (us-1) + "\t*: " + (us*2) + "\t/: " + (us/2) + "\t%: " + (us%2));

            Console.WriteLine("\nReference Type");
            Console.WriteLine("Object\t\tReference: " + o);
            Console.WriteLine("Dynamics:");
            Console.WriteLine("\t\tInteger: " + d_int);
            Console.WriteLine("\t\tCharacter: " + d_char);
            Console.WriteLine("\nString\t\tValue: " + str);
            Console.WriteLine("Classes\t\tName: " + p.getName());
        }
    }
}

/*
Source
https://www.tutorialspoint.com/csharp/csharp_data_types.htm
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/classes
https://docs.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/interfaces
*/